### PR TITLE
fix connection leak

### LINKED ISSUE #223: possible connection leak?
Yesterday I ran into emfile error when I disabled the VPN for sometime (mongodb server won't be available unless VPN is enabled). It sees to me like each connection failure leads to a tcp connection leak, as seen from the `port_count`. Let me know if this gives you any idea where the problem could be. I will try to debug further. I am using poolboy for connection pool

```
iex(42)> :erlang.system_info(:port_count)
158
iex(43)>
timestamp=2018-07-31T06:52:00.783Z level=error  message= Mongo.Protocol (#PID<0.829.0>) failed to connect: ** (Mongo.Error) auth failed for user usr_turbofin: Authentication failed. 18

timestamp=2018-07-31T06:52:01.381Z level=error  message= Mongo.Protocol (#PID<0.834.0>) failed to connect: ** (Mongo.Error) auth failed for user usr_turbofin: Authentication failed. 18

timestamp=2018-07-31T06:52:01.469Z level=error  message= Mongo.Protocol (#PID<0.831.0>) failed to connect: ** (Mongo.Error) auth failed for user usr_turbofin: Authentication failed. 18

nil
iex(44)> :erlang.system_info(:port_count)
161
iex(45)>
timestamp=2018-07-31T06:52:03.676Z level=error  message= Mongo.Protocol (#PID<0.835.0>) failed to connect: ** (Mongo.Error) auth failed for user usr_turbofin: Authentication failed. 18

nil
iex(46)> :erlang.system_info(:port_count)
timestamp=2018-07-31T06:52:07.882Z level=error  message= Mongo.Protocol (#PID<0.830.0>) failed to connect: ** (Mongo.Error) auth failed for user usr_turbofin: Authentication failed. 18

163
iex(47)>
timestamp=2018-07-31T06:52:08.794Z level=error  message= Mongo.Protocol (#PID<0.838.0>) failed to connect: ** (Mongo.Error) auth failed for user usr_turbofin: Authentication failed. 18

timestamp=2018-07-31T06:52:09.681Z level=error  message= Mongo.Protocol (#PID<0.837.0>) failed to connect: ** (Mongo.Error) auth failed for user usr_turbofin: Authentication failed. 18

nil
iex(48)>
timestamp=2018-07-31T06:52:11.518Z level=error  message= Mongo.Protocol (#PID<0.836.0>) failed to connect: ** (Mongo.Error) auth failed for user usr_turbofin: Authentication failed. 18

nil
iex(49)>
timestamp=2018-07-31T06:52:15.985Z level=error  message= Mongo.Protocol (#PID<0.832.0>) failed to connect: ** (Mongo.Error) auth failed for user usr_turbofin: Authentication failed. 18

timestamp=2018-07-31T06:52:19.305Z level=error  message= Mongo.Protocol (#PID<0.831.0>) failed to connect: ** (Mongo.Error) auth failed for user usr_turbofin: Authentication failed. 18

timestamp=2018-07-31T06:52:20.421Z level=error  message= Mongo.Protocol (#PID<0.834.0>) failed to connect: ** (Mongo.Error) auth failed for user usr_turbofin: Authentication failed. 18

timestamp=2018-07-31T06:52:20.477Z level=error  message= Mongo.Protocol (#PID<0.833.0>) failed to connect: ** (Mongo.Error) auth failed for user usr_turbofin: Authentication failed. 18

nil
iex(50)> :erlang.system_info(:port_count)
170
```

### PR DESCRIPTION
**Pull request by ananthakumaran**
fix #223

### CODE REVIEW

**Comment by ankhers** [comment]
There is no reason to have recon in this. Could you please remove it?

**Comment by ananthakumaran** [comment]
I can remove it if required. Please note that it's a test only dependency,
without it checking tcp connections count in the spec would get complicated.

**Comment by ankhers** [comment]
The function recon uses to count connections is fairly simple. You could just use

``` elixir
for port <- :erlang.ports(), {:name, 'tcp_inet'} == :erlang.port_info(port, :name), do: port
```

to get the tcp connections. I would rather not bring in a package for a single simple function.

**Comment by ananthakumaran** [comment]
thanks, removed recon.

File: lib/mongo/protocol.ex [hunk 1]
```diff
@@ -90,6 +93,7 @@ defmodule Mongo.Protocol do
       {:ok, ssl_sock} ->
         {:ok, %{s | socket: {:ssl, ssl_sock}}}
       {:error, reason} ->
+        :gen_tcp.close(sock)
```
Comment (ankhers) [inline lib/mongo/protocol.ex:96]: Is it guaranteed to be a tcp socket at this point? The returned error has the `:ssl` tag on it.

Comment (ananthakumaran) [reply lib/mongo/protocol.ex:96]:
```
iex(1)> {:ok, sock} = :gen_tcp.connect('www.google.com', 80, [])
{:ok, #Port<0.3165>}
iex(2)> :ssl.connect(sock, [])
{:error, :ssl_not_started}
iex(3)> Enum.count(:erlang.ports())
7
iex(4)> :gen_tcp.close(sock)
:ok
iex(5)> Enum.count(:erlang.ports())
6
```
I can't find anything in the ssl:connect method doc, so I am not 100% sure about all cases. ~~But it seems like the passed sock was not closed from what I have tested so far~~.

Comment (ananthakumaran) [reply lib/mongo/protocol.ex:96]: Looks like the ssl app was not started in the above case, I experimented some more, most of the time, the socket seems to get closed on error. But calling `gen_tcp.close` never raises any error, even if the port is already closed.

**Comment by ankhers** [comment]
Thanks!

**Pull request merged**